apps/heft/src/utilities/WatchFileSystemAdapter.ts

```diff
@@ -0,0 +1,257 @@
+   * Returns an object containing data about the state of said file (or folder).
+   */
+  getStateAndTrackAsync(filePath: string): Promise<IWatchedFileState>;
+}
+
+interface ITimeEntry {
+  timestamp: number;
+  safeTime: number;
+}
+
+/**
+ * A filesystem adapter for use with the "fast-glob" package. This adapter tracks file system accesses
+ * to initialize `watchpack`.
+ */
+export class WatchFileSystemAdapter implements IWatchFileSystemAdapter {
+  private _files: Map<string, number> = new Map();
+  private _contexts: Map<string, number> = new Map();
+  private _missing: Map<string, number> = new Map();
+
+  private _watcher: Watchpack | undefined;
+
+  private _lastFiles: Map<string, number> | undefined;
+  private _lastQueryTime: number | undefined;
+  private _times: Map<string, ITimeEntry> | undefined;
+
+  /** { @inheritdoc fs.readdirSync } */
+  public readdirSync: ReaddirSynchronousMethod = ((filePath: string, options?: IReaddirOptions) => {
+    filePath = path.normalize(filePath);
+
+    try {
+      if (options?.withFileTypes) {
+        const results: fs.Dirent[] = fs.readdirSync(filePath, options);
+        this._contexts.set(filePath, Date.now());
+        return results;
+      } else {
+        const results: string[] = fs.readdirSync(filePath);
+        this._contexts.set(filePath, Date.now());
+        return results;
+      }
+    } catch (err) {
+      this._missing.set(filePath, Date.now());
+      throw err;
+    }
+  }) as ReaddirSynchronousMethod;
+
+  /** { @inheritdoc fs.readdir } */
+  public readdir: ReaddirAsynchronousMethod = (
+    filePath: string,
+    optionsOrCallback: IReaddirOptions | ReaddirStringCallback,
+    callback?: ReaddirDirentCallback | ReaddirStringCallback
+  ) => {
+    filePath = path.normalize(filePath);
+    // Default to no options, which will return a string callback
+    let options: IReaddirOptions | undefined;
+    if (typeof optionsOrCallback === 'object') {
+      options = optionsOrCallback;
+    } else if (typeof optionsOrCallback === 'function') {
+      callback = optionsOrCallback;
+    }
+
+    if (options?.withFileTypes) {
+      fs.readdir(filePath, options, (err: NodeJS.ErrnoException | null, entries: fs.Dirent[]) => {
+        if (err) {
+          this._missing.set(filePath, Date.now());
+        } else {
+          this._contexts.set(filePath, Date.now());
+        }
+        (callback as ReaddirDirentCallback)(err, entries);
+      });
+    } else {
+      fs.readdir(filePath, (err: NodeJS.ErrnoException | null, entries: string[]) => {
+        if (err) {
+          this._missing.set(filePath, Date.now());
+        } else {
+          this._contexts.set(filePath, Date.now());
+        }
+        (callback as ReaddirStringCallback)(err, entries);
+      });
+    }
+  };
+
+  /** { @inheritdoc fs.lstat } */
+  public lstat: StatAsynchronousMethod = (filePath: string, callback: StatCallback): void => {
```

Comment:
```suggestion
  public lstat: typeof fs.lstat = (filePath: string, callback: StatCallback): void => {
```

To get rid of the dependency on `@nodelib/*`

Reply:
Using `typeof fs.lstat` doesn't work due to function overload typings.